flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java

```diff
@@ -0,0 +1,222 @@
+                            SqsExceptionClassifiers.getNotAuthorizedExceptionClassifier(),
+                            getSdkClientMisconfiguredExceptionClassifier()));
+
+    private final Counter numRecordsOutErrorsCounter;
+
+    /* Url of SQS */
+    private final String sqsUrl;
+
+    /* The sink writer metric group */
+    private final SinkWriterMetricGroup metrics;
+
+    /* Flag to whether fatally fail any time we encounter an exception when persisting records */
+    private final boolean failOnError;
+
+    SqsSinkWriter(
+            ElementConverter<InputT, SendMessageBatchRequestEntry> elementConverter,
+            Sink.InitContext context,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            boolean failOnError,
+            String sqsUrl,
+            SdkClientProvider<SqsAsyncClient> clientProvider,
+            Collection<BufferedRequestState<SendMessageBatchRequestEntry>> initialStates) {
+        super(
+                elementConverter,
+                context,
+                AsyncSinkWriterConfiguration.builder()
+                        .setMaxBatchSize(maxBatchSize)
+                        .setMaxBatchSizeInBytes(maxBatchSizeInBytes)
+                        .setMaxInFlightRequests(maxInFlightRequests)
+                        .setMaxBufferedRequests(maxBufferedRequests)
+                        .setMaxTimeInBufferMS(maxTimeInBufferMS)
+                        .setMaxRecordSizeInBytes(maxRecordSizeInBytes)
+                        .build(),
+                initialStates);
+        this.failOnError = failOnError;
+        this.sqsUrl = sqsUrl;
+        this.metrics = context.metricGroup();
+        this.numRecordsOutErrorsCounter = metrics.getNumRecordsOutErrorsCounter();
+        this.clientProvider = clientProvider;
+    }
+
+    @Override
+    protected void submitRequestEntries(
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        final SendMessageBatchRequest batchRequest =
+                SendMessageBatchRequest.builder().entries(requestEntries).queueUrl(sqsUrl).build();
+
+        CompletableFuture<SendMessageBatchResponse> future =
+                clientProvider.getClient().sendMessageBatch(batchRequest);
+
+        future.whenComplete(
+                (response, err) -> {
+                    if (err != null) {
+                        handleFullyFailedRequest(err, requestEntries, requestResult);
+                    } else if (response.failed() != null && response.failed().size() > 0) {
+                        handlePartiallyFailedRequest(response, requestEntries, requestResult);
+                    } else {
+                        requestResult.accept(Collections.emptyList());
+                    }
+                });
```

Comment:
What happens when `handleFullyFailedRequest` throws an `Exception`? Could we handle this?

```
future.whenComplete(...)
         .exceptionally(ex -> getFatalExceptionCons().accept(ex));
```

Reply:
@hlteoh37 My understanding here was that on Exception, whenComplete will put the exception under err object in (response, err) and then it will be handled under if (err != null) block? I can see the same implementation for all the other sinks

Reply:
Yep - my concern is more around if the block within `whenComplete(...)` throws an exception. We've found in other sinks that this leads to a "deadlock" like situation because we don't have `.exceptionally(...)` for the future.

Reply:
Understood, Added the exceptionally block